crates/router/src/core/user.rs

```diff
@@ -2760,3 +2760,424 @@ pub async fn list_profiles_for_user_in_org_and_merchant_account(
+    user_from_token: auth::UserFromToken,
+) -> UserResponse<user_api::TokenResponse> {
+    if user_from_token.merchant_id == request.merchant_id {
+        return Err(UserErrors::InvalidRoleOperationWithMessage(
+            "User switching to same merchant".to_string(),
+        )
+        .into());
+    }
+
+    let user = user_from_token.get_user_from_db(&state).await?;
+    let key_manager_state = &(&state).into();
+    let role_info = roles::RoleInfo::from_role_id(
+        &state,
+        &user_from_token.role_id,
+        &user_from_token.merchant_id,
+        &user_from_token.org_id,
+    )
+    .await
+    .change_context(UserErrors::InternalServerError)?;
+
+    let (org_id, merchant_id, profile_id, role_id) = match role_info.get_entity_type() {
+        EntityType::Internal => {
+            let merchant_key_store = state
+                .store
+                .get_merchant_key_store_by_merchant_id(
+                    key_manager_state,
+                    &request.merchant_id,
+                    &state.store.get_master_key().to_vec().into(),
+                )
+                .await
+                .to_not_found_response(UserErrors::MerchantIdNotFound)?;
+
+            let merchant_account = state
+                .store
+                .find_merchant_account_by_merchant_id(
+                    key_manager_state,
+                    &request.merchant_id,
+                    &merchant_key_store,
+                )
+                .await
+                .to_not_found_response(UserErrors::MerchantIdNotFound)?;
+
+            let profile_id = state
+                .store
+                .list_business_profile_by_merchant_id(
+                    key_manager_state,
+                    &merchant_key_store,
+                    &request.merchant_id,
+                )
+                .await
+                .change_context(UserErrors::InternalServerError)?
+                .first()
+                .ok_or(UserErrors::InternalServerError)?
+                .profile_id
+                .clone();
+
+            (
+                merchant_account.organization_id,
+                request.merchant_id,
+                profile_id,
+                user_from_token.role_id.clone(),
+            )
+        }
+
+        EntityType::Organization => {
+            let merchant_id = state
+                .store
+                .list_merchant_accounts_by_organization_id(
+                    key_manager_state,
+                    user_from_token.org_id.get_string_repr(),
+                )
+                .await
+                .change_context(UserErrors::InternalServerError)?
+                .into_iter()
+                .find(|account| *account.get_id() == request.merchant_id)
+                .ok_or(UserErrors::InvalidRoleOperationWithMessage(
+                    "No such merchant id found for the user in the org".to_string(),
+                ))?
+                .get_id()
+                .clone();
```

Comment:
Get the `merchant_account` for the `req.merchant_id` directly and check the org_id.